### PR TITLE
[stable-1.80]  fix(vendor): Strip excluded build targets 

### PR DESCRIPTION
**Pull request by epage**
Stable backports

* #14368

In order to make CI pass, the following PRs are also cherry-picked:

### CODE REVIEW

File: src/cargo/ops/vendor.rs [hunk 1]
```diff
@@ -220,7 +220,7 @@ fn sync(
         let pathsource = PathSource::new(src, id.source_id(), gctx);
```
Comment (epage) [inline src/cargo/ops/vendor.rs:220]: this is targeting `rust-1.80.0`.  Did that as a placeholder until we have `rust-1.80.2`

Comment (weihanglo) [reply src/cargo/ops/vendor.rs:220]: That branch is for `1.80` and would never change as far as I can tell.

**Review by weihanglo — APPROVED** [review]

**Comment by epage** [comment]
> [weihanglo](https://github.com/weihanglo) approved these changes

FYI, this intentionallty is broken.  I wanted to get this up for the conversation but I still need to work through the build breakages from other parts of cargo being different and then figure out what unblocking backports are needed.

**Comment by weihanglo** [comment]
@rfcbot fcp merge

I propose to backport #14368 

There is a bug in 1.80.0/1.80.1 that `cargo vendor` generates broken unbuildable vendored Git sources when

* The git dependency contains a build script
* The build script is explicitly excluded via `package.exclude`.

**Comment by epage** [comment]
FYI I've given a heads up to the release team at https://rust-lang.zulipchat.com/#narrow/stream/241545-t-release/topic/1.2E80.2E2

**Comment by weihanglo** [comment]
Should we bump the version of cargo crate to 0.81.1?

**Comment by epage** [comment]
According to @pietroalbini, this doesn't meet the bar for a new point release, only piggy-backing off of an existing point release, as this is limited to `cargo vendor` users and can be worked around by running `cargo vendor` on a different version.

This aligns with our last `cargo vendor` breakage (#11192).  That was broken in 1.64 and we didn't fix it until 1.68 (#11414).  It doesn't look like we even did a beta backport for that.

**Comment by weihanglo** [comment]
> This aligns with our last `cargo vendor` breakage (#11192). That was broken in 1.64 and we didn't fix it until 1.68 (#11414). It doesn't look like we even did a beta backport for that.


I have no strong inclination for stable backport after you pointed this out.

**Pull request closed without merging**